compiler/packages/babel-plugin-react-compiler/src/Utils/DisjointSet.ts

```diff
@@ -124,6 +124,42 @@ export default class DisjointSet<T> {
     return [...sets.values()];
   }
 
+  copy(): DisjointSet<T> {
+    const copy = new DisjointSet<T>();
+    copy.#entries = new Map(this.#entries);
+    return copy;
+  }
+
+  equals(other: DisjointSet<T>): boolean {
+    if (this.size !== other.size) {
+      return false;
+    }
+    const rootMap = new Map<T, T>();
+    for (const item of this.#entries.keys()) {
+      const otherRoot = other.find(item);
+      if (otherRoot === null) {
+        return false;
+      }
+      const thisRoot = this.find(item);
+      CompilerError.invariant(thisRoot != null, {
+        reason: 'Expected item to be in set',
+        loc: null,
+      });
+      if (rootMap.get(thisRoot) == null) {
+        rootMap.set(thisRoot, otherRoot);
+      } else if (rootMap.get(thisRoot) !== otherRoot) {
+        return false;
+      }
+    }
+    for (const item of other.#entries.keys()) {
+      const thisRoot = this.find(item);
+      if (thisRoot === null) {
+        return false;
+      }
+    }
+    return true;
+  }
```

Comment:
can you add some tests for this?

imagine two Sets:
1. `this` has groups `{a}`, `{b}`
2. `other` has one group, `{a, b}`

we'd iterate this:
* `a`. otherRoot = a or b, thisRoot = a, rootMap doesn't have a so we continue
* `b`. otherRoot = a or b, thisRoot = b, rootMap doesn't have b so we continue

now we iterate other:
* `a`, thisRoot = a which is non-null so continue
* `b`, thisRoot = b which is non-null so continue

So i think this example ends up incorrectly returning true (equals) when they're not equal.